GFS components update for hpc-stack support

KateFriedman-NOAA opened the pull request:
Update system components to versions that now support hpc-stack:

- ufs-weather-model @ [9350745](https://github.com/ufs-community/ufs-weather-model/tree/9350745855aebe0790813e0ed2ba5ad680e3f75c)
- GSI @ [9c1fc15](https://github.com/NOAA-EMC/GSI/tree/9c1fc15d42573b398037319bbf8d5143ad126fb6)
- GLDAS @ [gldas_gfsv16_release.v1.15.0](https://github.com/NOAA-EMC/GLDAS/releases/tag/gldas_gfsv16_release.v1.15.0)
- UFS_UTILS @ ufs_utils_1_4_0 (already in develop)
- EMC_post @ [ff42e02](https://github.com/NOAA-EMC/EMC_post/tree/ff42e0227d6100285d4179a2572b700fd5a959cb)
- EMC_gfs_wafs @ [c2a29a6](https://github.com/NOAA-EMC/EMC_gfs_wafs/tree/c2a29a67d9432b4d6fba99eac7797b81d05202b6)
- TC_tracker @ [TC_tracker.v1.1.15.4](https://github.com/NOAA-EMC/TC_tracker/releases/tag/TC_tracker.v1.1.15.4)

EMC_verif-global version not updating in this PR.

Workflow updates:

- update FIX_DIR paths in link_fv3gfs.sh to point to developmental set instead of frozen GFSv16 set (fix_nco_gfsv16 -> fix)
- update module loads to use hpc-stack (hpc/1.1.0) on supported platforms
- update obsproc pointers to versions on disk updated for hpc-stack
- set gfs_util build to run on all machines (stack build update)
- update build and link scripts to ingest component build/link step updates
- update build_fv3.sh to build CCPP version without waves (WW3 build updates not yet ready, disabling waves for now)
- ~~set DO_WAVE=NO by default for now~~ already committed with issue #178 PR
- ~~set RUN_CCPP=YES by default~~ already committed with issue #178 PR

Tested on WCOSS-Dell, Hera, and Orion. The feature/hpc-stack branch is also being used in port to Comet.

Close #164

KateFriedman-NOAA committed:
Issue #164 - update machine-setup.sh for Hera stack, update checkout/build/linking for UPP tag that supports hpc-stack, remove post patch in HERA.env

KateFriedman-NOAA committed:
Issue #164 - remove build_libs.sh, update module_base.hera for hpc-stack, plus updates for building enkf_chgres_recenter, gfs_bufr, and tocsbufr with hpc-stack

KateFriedman-NOAA committed:
Merge remote-tracking branch 'origin/develop' into feature/hpc-stack

* origin/develop:
  Issue #179 and issue #243 - update Fit2Obs to newm.1.2 tag and correct COMROOT path for Hera
  HOTFIX: Issue #241 - update EMC_verif-global tag to remove use of /tmp space

KateFriedman-NOAA committed:
Issue #164 - update GLDAS tag in Externals.cfg and stack updates for fbwndgfs and grib_util builds

KateFriedman-NOAA committed:
Issue #164 - updates to partial build and stack updates for tropcy build

KateFriedman-NOAA committed:
Issue #164 - stack updates for regrid_nemsio build

KateFriedman-NOAA committed:
Issue #164 - stack updates for UFS_UTILS build

KateFriedman-NOAA committed:
Updates for hpc-stack

- Remove unneeded ignores from .gitignore
- Update Externals.cfg and checkout.sh for component stack versions
- Update link_fv3gfs.sh for changed component links
- Further Hera module updates for hpc-stack

Refs: #164

KateFriedman-NOAA committed:
Merge remote-tracking branch 'origin/develop' into feature/hpc-stack

* origin/develop:
  remove flags that were not present in Makefiles
  update jasper to 2.x.25. remove compiler flag in fv3nc2nemsio
  Update EMC_verif-global tag to verif_global_v1.13.4
  another use of _d where an _4 is needed
  Adjust how target is set for build_workflow_utils
  Add workflow_utils to build_all and link scripts
  copy/paste error from enkf_chgres_recenter.fd to enkf_chgres_recenter_nc.fd.  When going from nemsio to netcdf, the linking of ip, sp and w3nco changed from _d to _4.
  update .gitignore to exclude build and install directories as well as compiled files. bugfix in build_workflow_utils.sh hack. nceplibs-ncio now creates the module ncio and not fv3gfs_ncio.
  add cmake build capability for workflow utilities
  Update Fit2Obs tag to newm.1.3 for bugfix
  Update Fit2Obs tag to newm.1.3 for bugfix
  Update EMC_verif-global tag for Hera bug
  Pull in nwpara/gfsv16.0.6 updates for parse-storm-type.pl
  HOTFIX: Update EMC_verif-global tag for Hera bug
  Issue #1 and issue #238 - update Externals.cfg to match checkout.sh updates for v16.0.6
  move errchk definition from script to job for wave prdgen and gempak
  adding definition of errchk which was undefined in these scripts
  Issue #1 and issue #233 - update GLDAS tag to gldas_gfsv16_release.v1.13.0
  Issue #1 and issue #241 - update EMC_verif-global tag
  Issue #1 and issue #238 - remove ak_10m grid from config.wave and update checkout.sh tags for ufs-weather-model and EMC_gfs_wafs
  Issue #1 and issue #226 - update UPP tag to upp_gfsv16_release.v1.1.3 in Externals.cfg
  Issue #1 and issue #226 - update UPP tag to upp_gfsv16_release.v1.1.3
  Issue #1 and issue #226 - update exgfs_atmos_grib2_special_npoess.sh for dbn_alert
  issue #227 reducing output bufr files to 64 levels per NCO request
  issue #227 reducing output bufr files to 64 levels per NCO request
  Issue #1:  update ecflow to be consistent with NCO's gfs.v16.0.4 and            update checkout to bring in new UPP tag (see issue #226)
  Upaded ush script scale_dec.sh
  Updated scripts gfs_v16.0

Conflicts:
	Externals.cfg
	sorc/checkout.sh

Refs: #164

KateFriedman-NOAA committed:
Updates from testing on Hera

- add WGRIB2 setenv to module_base.hera; wgrib2 was unknown in post jobs without
- set RUN_CCPP default to YES; should be running with CCPP now, consider removing variable

Refs: #164

KateFriedman-NOAA committed:
hpc-stack updates for WCOSS-Dell

- update OznMon, Radmon modulefiles
- update module_base.wcoss_dell_p3
- remove WW3 execs from link_fv3gfs.sh; will come from new build script
- update wcoss_dell_p3 section of machine-setup.sh for hpc-stack
- update checkout.sh for hpc-stack supported components (some test versions)

Refs: #164

KateFriedman-NOAA committed:
hpc-stack updates for Orion

- update OznMon and RadMon modulefiles
- update module_base.orion
- update Orion if-block in machine-setup.sh
- remove duplicate module use for WCOSS-Dell in machine-setup.sh

Refs: #164

KateFriedman-NOAA committed:
Merge remote-tracking branch 'origin/develop' into feature/hpc-stack

* origin/develop:
  Correct if-block syntax in config.postsnd and config.wavepostbndpnt after testing setup scripts
  FHMAX_WAV_IBP variable check for wavepostbndpnt job
  Bufr sounding job update
  Add config.resources.nco.static
  ecflow forecast job resource updates from NCO v16.0.7 install
  Update EMC_verif-global tag to verif_global_v1.13.5
  Add missing symlinks for WAFS source code folders
  Update EMC_gfs_wafs tag to gfs_wafs.v6.0.19
  Remove KEEPDATA from config.base.nco.static
  Pull in config changes from NCO v16.0.7 install
  Pull in workflow changes from NCO v16.0.7 install
  Update EMC_verif-global tag to verif_global_v1.13.4
  ecFlow resource adjustments from NCO for forecast and post jobs
  issue #257 shorten run time on Mars Dell1 file system

Refs: #164

KateFriedman-NOAA committed:
Remove UFS_UTILS symlinks no longer needed in link_fv3gfs.sh

Refs: #164

KateFriedman-NOAA committed:
Merge branch 'feature/hpc-stack' of https://github.com/NOAA-EMC/global-workflow into feature/hpc-stack

* 'feature/hpc-stack' of https://github.com/NOAA-EMC/global-workflow:
  Correct if-block syntax in config.postsnd and config.wavepostbndpnt after testing setup scripts
  FHMAX_WAV_IBP variable check for wavepostbndpnt job
  Bufr sounding job update
  Add config.resources.nco.static
  ecflow forecast job resource updates from NCO v16.0.7 install
  Update EMC_verif-global tag to verif_global_v1.13.5
  Add missing symlinks for WAFS source code folders
  Update EMC_gfs_wafs tag to gfs_wafs.v6.0.19
  Remove KEEPDATA from config.base.nco.static
  Pull in config changes from NCO v16.0.7 install
  Pull in workflow changes from NCO v16.0.7 install
  Update EMC_verif-global tag to verif_global_v1.13.4
  ecFlow resource adjustments from NCO for forecast and post jobs
  issue #257 shorten run time on Mars Dell1 file system

KateFriedman-NOAA committed:
Update GSI checkout to clone GSI-master @ 9c1fc15

Refs: #164

KateFriedman-NOAA committed:
Module base updates for WCOSS-Dell, Hera, and Orion

- add zlib module load
- reorder wgrib2, hdf5, and netcdf module loads to accommodate module dependencies

Refs: #164

KateFriedman-NOAA committed:
Merge remote-tracking branch 'origin/develop' into feature/hpc-stack

* origin/develop:
  Add two WAFS sorc folders to .gitignore
  Remove unneeded builds from build_all and partial build
  Delete unneeded files after workflow utils build was converted to cmake

Remove global_chgres exec symlink from link_fv3gfs.sh

Refs: #164

KateFriedman-NOAA committed:
Update OznMonBuild, RadMonBuild and module_base for hpc-stack on WCOSS-Dell

Refs: #164

KateFriedman-NOAA committed:
Remove Tide/Gyre reference from machine-setup.sh

Refs: #164

KateFriedman-NOAA committed:
Add wgrib2 module load to module_base.wcoss_dell_p3

- add wgrib2/2.0.8 module load
- add setting of WGRIB2 to equal wgrib2

Refs: #164

KateFriedman-NOAA committed:
Reduce npe_wav_gfs from 440 to 140 to dev users

Refs: #164

KateFriedman-NOAA committed:
Add support for separate gfs thread values

- nth_fv3_gfs was added to configs, add support to setup scripts for resource calculations

Refs: #164

KateFriedman-NOAA committed:
Update GLDAS and UFS_UTILS checkouts

- update GLDAS checkout to gldas_gfsv16_release.v1.15.0 tag; corrects build on Orion
- update UFS_UTILS checkout to 4f44bf89 hash of develop (supports stack)

Refs: #164

KateFriedman-NOAA committed:
Update Externals.cfg to match checkout.sh versions

Refs: #164

KateFriedman-NOAA committed:
Set WAFS checkout to be optional with -o flag

- add checkout_wafs to -o flag settings

Refs: #164

KateFriedman-NOAA committed:
Set DO_WAVE=NO in config.base

- set default DO_WAVE setting to NO in config.base.emc.dyn
- turn off waves until hpc-stack for WW3 execs build is ready

Refs: #164

KateFriedman-NOAA committed:
Update FIX_DIR paths in link_fv3gfs.sh

- change platform FIX_DIR paths from fix_nco_gfsv16 set to developmental fix set for v17+

Refs: #164

KateFriedman-NOAA committed:
Add separate gfs resource variables to fcst in config.resources

Refs: #164

KateFriedman-NOAA committed:
Adjust C768 fcst thread value for reduced node usage

- change C768 nth_fv3_gfs value from 7 to 2

Refs: #164

GeorgeGayno-NOAA commented:
@KateFriedman-NOAA The UFS_UTILS commit you are pointing is working. But be aware that is a branch in my fork. Let me see if I can get this merged. What is your deadline?

RussTreadon-NOAA commented:
hpc-stack entered the [NOAA-EMC/GSI](https://github.com/NOAA-EMC/GSI) [master](https://github.com/NOAA-EMC/GSI/tree/master) at [9c1fc15](https://github.com/NOAA-EMC/GSI/commit/9c1fc15d42573b398037319bbf8d5143ad126fb6).   The master has been updated several times since 9c1fc15.  Any NOAA-EMC/GSI master revision after the 9c1fc15 commit, completed  3/9/202, should work with hpc-stack.   @MichaelLueken-NOAA is the NOAA-EMC/GSI code manager.

WenMeng-NOAA commented:
@KateFriedman-NOAA That's EMC_post version should be working for FV3GFS offline post except wafs dataset generation. @YaliMao-NOAA  has been working on that issue.  I saw the version of  ufs-weather-model still uses old upp library 10.0.0. If @junwang-noaa provides you an up-to-date version, I might give you a new version of EMC_post for matching.

KateFriedman-NOAA commented:
> @KateFriedman-NOAA The UFS_UTILS commit you are pointing is working. But be aware that is a branch in my fork. Let me see if I can get this merged. What is your deadline?

Thank you for catching my mistake. That hash should have been 9d9dd23e37a51fcc5ff6b7499b834c85ab32e5f3 (develop branch with beta stack version). I also tested with your feature/stack branch. I'd like to submit this PR in the next week. Thanks!

KateFriedman-NOAA committed:
Update UFS_UTILS hash to 9d9dd23e37a51fcc5ff6b7499b834c85ab32e5f3

Refs: #164

KateFriedman-NOAA commented:
> @KateFriedman-NOAA That's EMC_post version should be working for FV3GFS offline post except wafs dataset generation. @YaliMao-NOAA has been working on that issue. 

@WenMeng-NOAA Gotcha...I'm also waiting for a version of WAFS that supports hpc-stack so I can stick with the EMC_post hash I have now and then update to a new EMC_post version when WAFS is ready.

> I saw the version of ufs-weather-model still uses old upp library 10.0.0. If @junwang-noaa provides you an up-to-date version, I might give you a new version of EMC_post for matching.

Okie dokie... @junwang-noaa do you recommend a newer hash of ufs-weather-model? Thanks!

GeorgeGayno-NOAA commented:
> > @KateFriedman-NOAA The UFS_UTILS commit you are pointing is working. But be aware that is a branch in my fork. Let me see if I can get this merged. What is your deadline?
> 
> Thank you for catching my mistake. That hash should have been 9d9dd23e37a51fcc5ff6b7499b834c85ab32e5f3 (develop branch with beta stack version). I also tested with your feature/stack branch. I'd like to submit this PR in the next week. Thanks!

Would you prefer UFS_UTILS use the 'official' stack? I can accelerate the merge.

KateFriedman-NOAA commented:
> Would you prefer UFS_UTILS use the 'official' stack? I can accelerate the merge.

Yes please, thanks @GeorgeGayno-NOAA !

WalterKolczynski-NOAA commented on parm/config/config.fv3:
```diff
@@ -110,9 +110,9 @@ case $case_in in
+        export npe_wav_gfs=140
         export nth_fv3=4
-        export nth_fv3_gfs=7
+        export nth_fv3_gfs=2
```
These seem like big resource changes.

KateFriedman-NOAA replied:
reverted

WalterKolczynski-NOAA commented on parm/config/config.resources:
```diff
@@ -168,8 +168,10 @@ elif [ $step = "fcst" ]; then
+    export nth_fcst_gfs=${nth_fv3_gfs:-2}
     if [[ "$CASE" == "C384" ]]; then nth_fcst=4; fi
     export npe_node_fcst=$(echo "$npe_node_max / $nth_fcst" | bc)
+    export npe_node_fcst_gfs=$(echo "$npe_node_max / $nth_fcst_gfs" | bc)
```
Why this second specification for resources?

KateFriedman-NOAA replied:
reverted

WalterKolczynski-NOAA commented on ush/rocoto/workflow_utils.py:
```diff
@@ -295,7 +295,10 @@ def get_resources(machine, cfg, task, reservation, cdump='gdas'):
+        if cdump in ['gfs'] and 'nth_%s_gfs' % task in cfg.keys():
+            threads = cfg['nth_%s_gfs' % ltask]
+        else:
+            threads = cfg['nth_%s' % ltask]
```
More with the second resource specification.

KateFriedman-NOAA replied:
reverted

WalterKolczynski-NOAA reviewed (changes requested):
Why are resource changes even included in a PR migrating to the stack?

KateFriedman-NOAA committed:
Reverting resource related changes

- will commit resource adjustments separately

Refs: #164

KateFriedman-NOAA committed:
Update TC_tracker version in config.vrfy

- new TC_tracker.v1.1.15.4 was made available by Jiayi Peng
- new version supports hpc-stack on supported platforms
- installed locally on WCOSS-Dells, Hera, and Orion

Refs: #164

KateFriedman-NOAA committed:
Update util_shared to 1.3.0 on WCOSS-Dell

- version changed in production for v16.0.9 footer fix

Refs: #164

KateFriedman-NOAA committed:
Component updates for UPP, WAFS, UFS_UTILS

- new tag in checkout for UFS_UTILS to use hpc/1.1.0
- new hash in checkout for UPP/EMC_post to support GTG and GTG code copy fix
- new hash in checkout for WAFS/EMC_gfs_wafs to support hpc-stack
- build script update for WAFS (cmake/stack)
- WAFS GCIP config update to fix runtime error

Refs: #164

KateFriedman-NOAA committed:
Update EMC_post, UFS_UTILS, EMC_gfs_wafs versions in Externals.cfg

Refs: #164

KateFriedman-NOAA committed:
ufs-weather-model hash update for WW3 support with hpc-stack

- update ufs-weather-model hash to 554aedcd63e4a7c5012570406132eaf76e249ca9
- update build_fv3.sh for new compile arguments for CCPP/WW3 build

Refs: #164

KateFriedman-NOAA committed:
Update WAFS execs to include .x

- update WAFS exec names in link_fv3gfs.sh to include ".x" extension that was added with new version

Refs: #164

KateFriedman-NOAA committed:
Add build script and modulefiles for WW3 build

- WW3 execs no longer built by FV3 build, need separate build now
- add new build_ww3prepost.sh script for building WW3 execs in ufs-weather-model checkout
- add modulefiles for WCOSS-Dell, Hera, and Orion

Refs: #164

KateFriedman-NOAA committed:
Change CCPP_SUITE from FV3_GFS_v16beta to FV3_GFS_v16

- newer ufs-weather-model hash supports FV3_GFS_v16 suite

Refs: #164

KateFriedman-NOAA committed:
Change fix file linking to scan for possible subfolders in set

- instead of providing subfolder names, scan for subfolders in FIX_DIR path
and then create symlinks based on existing subfolders in set
- will not have to keep updating list of subfolders now

Refs: #164

JessicaMeixner-NOAA committed:
update for ww3_pre/post build:
-- remove ww3 specific module files
-- updated module base files to be in sync with ufs-weather-model
   except that jasper is a more recent version which is needed for ww3_grib
-- added extra ww3 execs needed for this branch
-- added ww3prepost build to build_all and partial build

KateFriedman-NOAA committed:
Merge pull request #322 from JessicaMeixner-NOAA/hpcstackww3

Updates for ww3 pre and post exec builds for hpc-stack

KateFriedman-NOAA committed:
Sync merge with develop and WW3 build updates

- Merge remote-tracking branch 'origin/develop' into feature/hpc-stack
- WW3 build updates: revert back to separate module files for build;
  address modulefile reorganization in separate issue
- remove CCPP option with system build, doing CCPP by default now

Refs: #164

KateFriedman-NOAA committed:
Modulefile updates for testing with waves on

- updated ww3 and base modulefiles to add modules and update some
- added new: pio/2.5.2i, fms/2020.04.03
- updated esmf module to esmf/8_1_1 and nceppost/dceca26 to upp/10.0.6

Refs: #164

KateFriedman-NOAA committed:
Merge remote-tracking branch 'origin/develop' into feature/hpc-stack

Refs: #164

KateFriedman-NOAA committed:
Merge remote-tracking branch 'origin/develop' into feature/hpc-stack

KateFriedman-NOAA commented:
@junwang-noaa @yangfanglin @RussTreadon-NOAA @WenMeng-NOAA @YaliMao-NOAA  This PR is in pre-commit testing and is ready to go into global-workflow develop. Please confirm your respective component hash versions mentioned in [first comment of PR](https://github.com/NOAA-EMC/global-workflow/pull/316#issue-621133376) are still good or if there are more recent hashes/tags I should update to and do a final retest with. Thanks!

RussTreadon-NOAA commented:
NOAA-EMC/GSI at [9c1fc15](https://github.com/NOAA-EMC/GSI/tree/9c1fc15d42573b398037319bbf8d5143ad126fb6) is several months old (3/9/2021 commit).  The NOAA-EMC/GSI master was updated this morning (6/21/2021) at [a67b816](https://github.com/NOAA-EMC/GSI/commit/a67b816a448d23d2010e9c39741af65a1c56b8f1).   I assume developers will want to use the head of NOAA-EMC/GSI for development.   While this PR could be updated to a67b816, this commit will also become obsolete with time.   NOAA-EMC/GSI development continues moving forward.   Developers can always jump to the head of NOAA-EMC/GSI by editing their local copy of `sorc/checkout.sh`.   

My preference is to work with the most recent versions.   That said, if updating DA in this PR to a67b816 requires rerunning global parallels on multiple machines before this PR may be closed, this may not be a staff and compute cost we want to incur at present.  We may opt to close this PR with DA at 9c1fc15 and open new PRs to bring the workflow closer to the head of DA development.   

These comments may also apply to other workflow components (model, UPP, verification, etc)

WenMeng-NOAA commented:
@KateFriedman-NOAA I looked at the harsh version of EMC_post and upp lib version at ufs-weather-model (fv3gfs). The offline post and inline post are matched.

KateFriedman-NOAA commented:
> @KateFriedman-NOAA I looked at the harsh version of EMC_post and upp lib version at ufs-weather-model (fv3gfs). The offline post and inline post are matched.

Great, thanks for checking and confirming @WenMeng-NOAA !

KateFriedman-NOAA commented:
> NOAA-EMC/GSI at [9c1fc15](https://github.com/NOAA-EMC/GSI/tree/9c1fc15d42573b398037319bbf8d5143ad126fb6) is several months old (3/9/2021 commit). The NOAA-EMC/GSI master was updated this morning (6/21/2021) at [a67b816](https://github.com/NOAA-EMC/GSI/commit/a67b816a448d23d2010e9c39741af65a1c56b8f1). I assume developers will want to use the head of NOAA-EMC/GSI for development. While this PR could be updated to a67b816, this commit will also become obsolete with time. NOAA-EMC/GSI development continues moving forward. Developers can always jump to the head of NOAA-EMC/GSI by editing their local copy of `sorc/checkout.sh`.
> 
> My preference is to work with the most recent versions. That said, if updating DA in this PR to a67b816 requires rerunning global parallels on multiple machines before this PR may be closed, this may not be a staff and compute cost we want to incur at present. We may opt to close this PR with DA at 9c1fc15 and open new PRs to bring the workflow closer to the head of DA development.

Thanks @RussTreadon-NOAA , I agree with your points and will use the hash slatted for this commit but plan to roll in newer GSI hashes/tags in the near future. Users can update their versions manually for now if they want the latest GSI hashes. Updating single components moving forward will be easier than this initial update to hpc-stack versions.

KateFriedman-NOAA commented:
@WalterKolczynski-NOAA I believe your review changes have been resolved now. Please rereview and let me know of any other questions or requested changes. Thanks!
